### PR TITLE
migrate to new SetOrMapLiteral API

### PR DESCRIPTION
**Pull request by pq**
Migrates to new `SetOrMapLiteral` API.

A few notes on the unification:

* `always_specify_types` was missing the opportunity to flag set literals; this fixes that 👍 
* I added TODOs in places where existing logic was preserved but might better be widened to include Set literals too 🤔 
* I'm not sure what to do about the test for `node is MapLiteralEntry`:  will that interface go away too?

(A similar migration for `ForStatement2 ` to follow.)

/cc @stereotype441 @bwilkerson

### CODE REVIEW

File: lib/src/rules/prefer_const_literals_to_create_immutables.dart [hunk 1]
```diff
@@ -78,7 +78,12 @@ class _Visitor extends SimpleAstVisitor<void> {
   void visitListLiteral(ListLiteral node) => _visitTypedLiteral(node);
 
   @override
-  void visitMapLiteral(MapLiteral node) => _visitTypedLiteral(node);
+  void visitSetOrMapLiteral(SetOrMapLiteral node) {
+    // todo (pq): should this apply to set literals as well?
```
Comment (bwilkerson) [inline lib/src/rules/prefer_const_literals_to_create_immutables.dart]: I believe the answer is yes.

File: lib/src/rules/prefer_const_literals_to_create_immutables.dart [hunk 2]
```diff
@@ -113,7 +118,8 @@ class _Visitor extends SimpleAstVisitor<void> {
         (node is ParenthesizedExpression ||
             node is ArgumentList ||
             node is ListLiteral ||
-            node is MapLiteral ||
+            node is SetOrMapLiteral ||
+            // todo (pq): check if this needs updating with the SetOrMapLiteral unification
```
Comment (bwilkerson) [inline lib/src/rules/prefer_const_literals_to_create_immutables.dart]: No. We still use MapLiteralEntry to represent a `key : value` structure.

File: lib/src/rules/prefer_void_to_null.dart [hunk 3]
```diff
@@ -92,7 +92,10 @@ class _Visitor extends SimpleAstVisitor<void> {
         final literal = parent.parent;
         if (literal is ListLiteral && literal.elements.isEmpty) {
           return;
-        } else if (literal is MapLiteral && literal.entries.isEmpty) {
+        } else if (literal is SetOrMapLiteral &&
+            // todo (pq): should this apply to set literals as well?
```
Comment (bwilkerson) [inline lib/src/rules/prefer_void_to_null.dart]: Again, I think the answer is yes. Tests will confirm.

File: lib/src/rules/unnecessary_const.dart [hunk 4]
```diff
@@ -74,7 +74,12 @@ class _Visitor extends SimpleAstVisitor {
   visitListLiteral(ListLiteral node) => _visitTypedLiteral(node);
 
   @override
-  visitMapLiteral(MapLiteral node) => _visitTypedLiteral(node);
+  void visitSetOrMapLiteral(SetOrMapLiteral node) {
+    // todo (pq): should this apply to SetLiterals as well?
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_const.dart]: Yes.

File: lib/src/rules/prefer_const_literals_to_create_immutables.dart (hunk 2 above)
Comment (pq) [reply lib/src/rules/prefer_const_literals_to_create_immutables.dart]: Thanks!

**Comment by pq** [comment]
Thanks for the quick feedback @bwilkerson!

**Pull request merged**